WMF Framework/Event Platform Client/EventPlatformClientLibrary.swift

```diff
@@ -0,0 +1,605 @@
+ * - [mw:Wikimedia Product/Analytics Infrastructure/Event Platform Client](https://www.mediawiki.org/wiki/Wikimedia_Product/Analytics_Infrastructure/Event_Platform_Client)
+ * - [wikitech:Event Platform/Instrumentation How To](https://wikitech.wikimedia.org/wiki/Event_Platform/Instrumentation_How_To)
+ *
+ * **Note**: Events generated while offline are persisted between sessions until stream configuration has
+ * been downloaded and the events can be properly processed.
+ *
+ * ## Logging API
+ *
+ * - `log(stream, schema, version, data, domain?)` to log events; can be called before
+ *   configuration is available or even offline
+ * - `logger(schema, version)` to generate a shortcut to `log` for logging multiple events using the
+ *   same schema version, without having to specify those in every call
+ * - `logger(stream, schema, version)` to generate a shortcut to `log` for logging multiple
+ *   events using the same stream name and schema version, without having to specify those in every call
+ *
+ * ## Dependencies
+ *
+ * EPC relies on some functionality to be made available to it by the rest of the application. It depends on:
+ * - a `NetworkManager` which can `HTTP POST` requests and download data
+ * - a `StorageManager` which can persist data and recall or delete persisted data
+ *
+ * We describe the desired behaviors for both of those in EventPlatformClientProtocols.swift
+ */
+public class EPC {
+
+    // MARK: - Properties
+
+    public static let shared = EPC() // singleton
+
+    /**
+     * See [wikitech:Event Platform/EventGate](https://wikitech.wikimedia.org/wiki/Event_Platform/EventGate)
+     * for more information. Specifically, the section on **eventgate-analytics-external**.
+     */
+    private let EVENTGATE_URI: String
+    private let CONFIG_URI: String
+    /**
+     * `ENABLED` is a safeguard against logging events while the app is in background state.
+     */
+    private var ENABLED: Bool
+    private var CONFIG: [String: [String: Any]]? = nil
+    private var COPIED = [String: [String]]()
+    /**
+     * Updated with every `log` call, used for determining if session expired.
+     */
+    private var LAST_TS: Date = Date()
+    private var SESSION_ID: String? = nil
+    private var DEVICE_ID: String? = nil
+    private let ISO8601_FORMATTER: ISO8601DateFormatter
+
+    /**
+     * For persistent storage
+     */
+    private var storage_manager: StorageManager?
+
+    /**
+     * For handling HTTP requests
+     */
+    private var network_manager: NetworkManager?
+
+    /**
+     * Store events until the library is finished initializing
+     *
+     * The EPC library makes an HTTP request to a remote stream configuration service for information
+     * about how to evaluate incoming event data. Until this initialization is complete, we store any incoming
+     * events in this buffer.
+     */
+    private var input_buffer = [(stream: String, schema: String, data: [String: Any], domain: String?)]()
```

Comment:
I think we will need to take some extra efforts to make sure `input_buffer` and `cache` are thread-safe. There are multiple ways to do this - `ImageControllerCompletionManager.swift` has an example with it's `completions` and `tasks` properties. You make a serial dispatch queue property, then anytime you're reading or writing to `input_buffer` or `cache`, you wrap it up in a `queue.async`. Happy to answer any questions you have on this or pitch in if you run into trouble.

Reply:
Thanks for all your guidance and help getting this stuff multithread-safe!